### PR TITLE
Obey

### PR DESCRIPTION
**Pull request by Fluidbyte**
No more Joi in our lives...

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -27,7 +27,7 @@ utilized (`npm install modli-nedb --save`).
 
 ```javascript
 // Import all the Modli methods
-import { model, adapter, use, Joi } from 'modli';
+import { model, adapter, use } from 'modli';
```
Comment (stephen-ta) [inline README.md]: Should this have `obey` in it?

File: src/lib/model.js [hunk 2]
```diff
@@ -22,23 +22,28 @@ model.store = {};
...
-    model.store[m.name] = {};
+    model.store[m.name] = {}
   }
   // Build model object
-  let modelObj = {};
+  let modelObj = {}
   Object.keys(m).forEach((prop) => {
     if (prop !== 'version' && prop !== 'name') {
-      modelObj[prop] = m[prop];
+      if (prop === 'schema') {
```
Comment (TomFrost) [inline src/lib/model.js]: If we change this to `if (prop === 'schema' && !m.schema.validate) {`, that would set us set the schema to `obey.model(...)` in the model definition itself. The advantage there is that we'll likely be adding types/generators/modifiers to Obey _globally_ from some `lib/obey.js` file in each hub somewhere, and that will require importing that file in each model to be sure the generators and types and such have been created.

But if we import that file but don't _use_ it, we're looking at linting errors at best. At worst, someone could remove that line later, thinking it's not used, and break things without understanding why.  So this would give us a sane way to use it.

(Actually, looking through the code, there's no part of Modli that actually cares what the schema _is_ beyond this line. Technically, with this change, we could allow people to plug in any validation lib they want as long as it exposes a validate() method. Totally not worth a focus on supporting right now though)

File: README.md (hunk 1 above)
Comment (TomFrost) [inline README.md]: Joi was here before because you actually needed it to define the schema. Obey's schema is defined as an object literal -- you don't need obey to construct it. lib/model.js uses Obey on your behalf to do that.

**Comment by TomFrost** [comment]
:accept: apart from my comment! With the caveat that the changes were hard to follow because of semis being removed, so I'm relying on passing tests here mostly ;-)

File: src/lib/model.js [hunk 3]
```diff
@@ -48,38 +53,34 @@ model.add = (m) => {
...
+      const v = version || this.defaultVersion
       // Return validation
-      return Joi.validate(data, Joi.object().keys(this.schemas[v].schema), (err) => {
-        if (err) {
-          return model.formatValidationError(err);
-        }
-        return null;
-      });
+      return this.schemas[v].schema.validate(data)
+        .catch(err => model.formatValidationError(err.collection))
```
Comment (stephen-ta) [inline src/lib/model.js]: So it looks like this might be grabbing the error, and then reformatting it? Does it also return it to the end user so they can _catch_ it within their app?

Comment (stephen-ta) [inline src/lib/model.js]: Nvm, I see the function `formatValidationError` later on in the code, and it throws!

File: test/index.spec.js [hunk 4]
```diff
@@ -1,8 +1,8 @@
 /* eslint no-unused-expressions: 0 */
 /* global expect, request, describe, it, before, after */
-import './setup';
-import { model, adapter, use, Joi } from '../src/index';
-import nedb from 'modli-nedb';
+import './setup'
+import { model, adapter, use } from '../src/index'
```
Comment (stephen-ta) [inline test/index.spec.js]: So I mentioned this earlier too about `obey` not being imported here. Is that because it's not being called within this file, but the schema gets passed to _obey_ somewhere else behind the scenes?

File: test/src/lib/model.spec.js [hunk 5]
```diff
@@ -10,125 +10,141 @@ describe('model', () => {
...
+        .then(data => {
+          expect(data).to.be.an.object
+        })
+    })
     // Fail validation condition
     it('fails validation when object does not match rules', () => {
-      const failTest = testModel.validate(testFailDataV2);
-      expect(failTest).to.not.be.null;
-    });
+      return testModel.validate(testFailDataV2)
```
Comment (stephen-ta) [inline test/src/lib/model.spec.js]: If this test doesn't _throw_ but actually has valid data, then this test isn't currently checking for that in order for the test to fail and alert you. It's only setup to expect a catch.

File: src/lib/model.js (hunk 2 above)
Comment (Fluidbyte) [inline src/lib/model.js]: Good call, changed.

Comment (TomFrost) [inline src/lib/model.js]: Bug in my suggestion -- if someone defines a `validate` key property, that fails.  But we could do this safely:
```
if (prop === 'schema' && typeof m.schema.validate !== 'function') {
```
Or, if you have lodash in here, _.isFunction would be better... JS likes to change the typeof to object is someone specifies a property on a function.

File: test/src/lib/model.spec.js [hunk 6]
```diff
@@ -10,125 +10,141 @@ describe('model', () => {
...
       // Define custom formatter
       model.customValidationError = (err) => {
-        return err.details[0].message;
-      };
-      const testCustom = testModel.validate(testFailDataV2);
-      expect(testCustom).to.equal('"id" must be a number');
-    });
+        return err[0].message
+      }
+      return testModel.validate(testFailDataV2)
```
Comment (stephen-ta) [inline test/src/lib/model.spec.js]: Same `.catch()` only. Might be a couple more. I will let you look through instead of commenting on all of them.

File: test/index.spec.js (hunk 4 above)
Comment (Fluidbyte) [inline test/index.spec.js]: Yes, the model gets applied to the schema if not already set.

File: test/src/lib/model.spec.js (hunk 5 above)
Comment (TomFrost) [inline test/src/lib/model.spec.js]: Frigging great catch ^

File: test/src/lib/model.spec.js [hunk 7]
```diff
@@ -10,125 +10,141 @@ describe('model', () => {
...
+        })
+    })
     // Fail on older version
     it('fails validation on older version of schema', () => {
-      const failTest = testModel.validate(testFailDataV1, 1);
-      expect(failTest).to.not.be.null;
-    });
-  });
+      return testModel.validate(testFailDataV1, 1)
+        .catch(err => {
```
Comment (stephen-ta) [inline test/src/lib/model.spec.js]: Actually, this might be the last one :)

**Comment by TomFrost** [comment]
@Fluidbyte I amended my suggestion while you were committing it D:. Here it is again since Github's hidden it:

Bug in my suggestion -- if someone defines a `validate` key property, that fails.  But we could do this safely:
```
if (prop === 'schema' && typeof m.schema.validate !== 'function') {
```
Or, if you have lodash in here, _.isFunction would be better... JS likes to change the typeof to object is someone specifies a property on a function.

**Comment by stephen-ta** [comment]
Good job Kent!

**Comment by Fluidbyte** [comment]
Ok, think I got everything.

**Comment by Fluidbyte** [comment]
Agreed, but rewriting tests needs to occur on a much larger level (a few of the tests are inter-dependent right now), so I'm saving a refactor of all that until after we get this working and integrated internally.

**Comment by TomFrost** [comment]
Fair! :accept:

**Comment by stephen-ta** [comment]
No devlab yet -- I had to switch over to modli just now to dig through the code for something and just noticed my favorite test suite command does not work :(

**Comment by Fluidbyte** [comment]
Yeah, devlab will be another addition, it wasn't _needed_ for this as I'm only using a local adapter and the db is in-memory.
